golang/cosmos/types/address_hooks.go

```diff
@@ -177,41 +177,67 @@ func extractBaseTransferData(transferData transfertypes.FungibleTokenPacketData,
 // If newPacket is not nil, it is populated with a new transfer packet whose
 // corresponding Sender or Receiver is replaced with the extracted base address.
 func ExtractBaseAddressFromPacket(cdc codec.Codec, packet ibcexported.PacketI, role AddressRole, newPacket *channeltypes.Packet) (string, error) {
+	var newDataP *[]byte
+	if newPacket != nil {
+		// Capture the data for the new packet.
+		newDataP = new([]byte)
+	}
+	target, err := ExtractBaseAddressFromData(cdc, packet.GetData(), role, newDataP)
+	if err != nil {
+		return target, err
+	}
+
+	if newPacket == nil {
+		return target, nil
+	}
+
+	// Create the new packet.
+	th := packet.GetTimeoutHeight()
+	*newPacket = channeltypes.NewPacket(
+		*newDataP, packet.GetSequence(),
+		packet.GetSourcePort(), packet.GetSourceChannel(),
+		packet.GetDestPort(), packet.GetDestChannel(),
+		clienttypes.NewHeight(th.GetRevisionNumber(), th.GetRevisionHeight()),
+		packet.GetTimeoutTimestamp(),
+	)
+
+	return target, nil
+}
+
+// ExtractBaseAddressFromData returns the base address from a transfer packet's data,
+// either Sender (if role is RoleSender) or Receiver (if role is RoleReceiver).
+// Errors in determining the base address are ignored... we then assume the base
+// address is exactly the original address.
+// If newDataP is not nil, it is populated with new transfer packet data whose
+// corresponding Sender or Receiver is replaced with the extracted base address.
+func ExtractBaseAddressFromData(cdc codec.Codec, data []byte, role AddressRole, newDataP *[]byte) (string, error) {
 	transferData := transfertypes.FungibleTokenPacketData{}
-	if err := cdc.UnmarshalJSON(packet.GetData(), &transferData); err != nil {
+
+	if err := cdc.UnmarshalJSON(data, &transferData); err != nil {
 		return "", err
 	}
 
 	var newTransferData *transfertypes.FungibleTokenPacketData
-	if newPacket != nil {
-		// Capture the transfer data for the new packet.
+	if newDataP != nil {
+		// Capture the transfer data for the new packet data.
 		newTransferData = &transfertypes.FungibleTokenPacketData{}
 	}
 	target, err := extractBaseTransferData(transferData, role, newTransferData)
 	if err != nil {
 		return target, err
 	}
 
-	if newPacket == nil {
+	if newDataP == nil {
 		return target, nil
 	}
 
 	// Create a new packet with the new transfer packet data.
 	// Re-serialize the packet data with the base addresses.
-	newData, err := cdc.MarshalJSON(newTransferData)
+	bz, err := cdc.MarshalJSON(newTransferData)
```

Comment:
Should we switch to `GetBytes` for the actual runtime? That should reduce the number of times where we have to store the packet data as sending chains are likely to use `GetBytes` themselves.

Also I just realize the previous PR (#10852) was cloning the packet data if the base addr and original address matched. Any reason to have dropped that?